### PR TITLE
Some more suggestions

### PR DESCRIPTION
**Pull request by martindurant**
This turned into a bit of a rewrite as I made tests that would pass.

### CODE REVIEW

**Comment by martindurant** [comment]
I guess if you were to merge from master you would get a sensible diff! The only important parts are in fsspec/implementations/reference.py

File: fsspec/implementations/reference.py [hunk 1]
```diff
@@ -119,63 +92,47 @@ def open_refs(path):
...
-        return self.zmetadata["metadata"][key]
+        if key in self._items:
+            return self._items[key]
+        elif key in self.zmetadata:
+            return json.dumps(self.zmetadata[key]).encode()
+        field, sub_key = key.split("/")
+        # Chunk keys can be loaded from row group and cached in LRU cache
+        record, ri, chunk_size = self._key_to_record(key)
+        if chunk_size == 0:
+            # raise KeyError ?
```
Comment (agoodm) [inline fsspec/implementations/reference.py]: This check was added because some refs I have tested would store variables consisting of a single empty string as a 0 length zarr array instead of as a global attribute.

File: fsspec/implementations/reference.py [hunk 2]
```diff
@@ -53,32 +51,6 @@ def _protocol_groups(paths, references):
     return out
 
 
-class RefsValuesView(collections.abc.ValuesView):
```
Comment (agoodm) [inline fsspec/implementations/reference.py:56]: I personally think it would be very worthwhile to include these changes because they make dramatic differences in performance when iterating over the entire reference set sequentially. I'll explain why in more detail in a bit, but just to give you an idea of the significance of this it takes nearly 40 minutes for me to iterate over the entire NWM reference set (which is a lot slower than the original JSON, even after including the load time for that). through the lazy ref mapper with the default implementation of `values()` and `items()`. When overriding with these implementations, it takes just 2 minutes, so nearly a 20x improvement.

The main reason why is the default implementations of `values()` and `items()` call `__getitem__()` at every iteration, and this has a ton of overhead compared to retrieving a key from a normal dictionary. In particular it has to:

- Convert the string chunk key to a 1D integer index
- Figure out which parquet file to load from and from which row
- Check whether each item is a raw or url
- Additionally if the urls are categorical indexing into it is also a lot slower.

These are all fine for random access since the amount of time it will take to load the actual chunk will be much greater than the time it takes to load each reference, but for situations where we just need to process all references and not actually load the data (perhaps in combine?), this adds up really fast.

So basically, this custom implementation takes advantage of the fact that we know the correct order of the chunk keys in advance, we can just iterate them separately rather than explicitly calling `__getitem__()` so we can easily eliminate the first two bottlenecks from having to process each key. And since each iteration is done over each parquet file, we can greatly reduce the number of times we need to check for whether the reference is a raw or url since this can be checked once per file if there are less than 4 columns. Finally categorical URL arrays are fully expanded so they can be indexed into much faster. In this way the four bottlenecks I mentioned above are eliminated.

File: fsspec/implementations/reference.py [hunk 3]
```diff
@@ -190,68 +147,31 @@ def _key_to_record(self, key):
...
 
     def __delitem__(self, key):
-        del self._items[key]
+        if key in self._items:
+            del self._items[key]
+        elif key in self.zmetadata:
+            del self.zmetadata[key]
+        else:
+            # set appropriate reference to zeros in memory
+            raise KeyError
```
Comment (agoodm) [inline fsspec/implementations/reference.py]: I didn't give this part too much thought when implementing it, but I am wondering how this would work, mainly because it's possible for references to be purged from memory anyway through the LRU cache, so if a delete operation set them to null, they could end up being reset to their original values again later unknowingly which would make this op inconsistent. Unless you know of a use-case where it is necessary to be able to delete or null refs lazily, I think just defaulting to a KeyError here is fine. Could also set `self._items[key]` to `None`?

**Review by agoodm — COMMENTED** [review]
Thanks for looking this over, I have made a few comments, the main important one being about the iteration optimizations I made.

File: fsspec/implementations/reference.py (hunk 3 above)
Comment (martindurant) [reply fsspec/implementations/reference.py]: You are definitely right that deleting real non-metadata references seems unlikely, so we could defer this. However, it would be nice to have editable reference sets, as the current implementation is, so that you can in theory change out parts of your data ( http://martindurant.github.io/blog/mutable-kerchunk/ ). In that case, we need to keep track of all amended/deleted references, and indeed, I would I suppose use self._items to do that. Since the current lookup looks there first, it will have priority over the original, and we just have to check for None (which would raise KeyError on access).

File: fsspec/implementations/reference.py (hunk 2 above)
Comment (martindurant) [reply fsspec/implementations/reference.py:56]: You make some very good points! I guess this would be a critical failure when combining if it takes so long. Would you like me to undo these changes, or will you?

File: fsspec/implementations/reference.py (hunk 1 above)
Comment (martindurant) [reply fsspec/implementations/reference.py]: A length-zero array has no chunk file but does have attributes?
So not a KeyError, please ignore.

File: fsspec/implementations/reference.py (hunk 2 above)
Comment (agoodm) [reply fsspec/implementations/reference.py:56]: Great! I can go ahead and do this.

File: fsspec/implementations/reference.py (hunk 1 above)
Comment (agoodm) [reply fsspec/implementations/reference.py]: Yep!

File: fsspec/implementations/reference.py (hunk 3 above)
Comment (agoodm) [reply fsspec/implementations/reference.py]: Thanks for sharing that, now I see what you are getting at. So for now we could do:

- Set `self._item[key] = None`
- Raise KeyError in `__getitem__`  if  `self._item[key] is None`

Then I suppose later follow-up work would be to implement a method to RefFS that writes the updated references to the parquet files?

Comment (martindurant) [reply fsspec/implementations/reference.py]: Exactly. We should even know exactly which files need rewriting, the rest to be left in place or copied if the new version is elsewhere. This will be great for appending.
